File: year/2025/day02/task2/main.py
"""
Author: Jose A. Romero
Puzzle: Advent of Code (year=2025 ; day=2 ; task=2)
"""

import sys
from itertools import accumulate
from bisect import bisect_left


def main() -> None:
    invalid_ids_set = set()
    for i in range(1, 10**6):
        str_i = str(i)

        for r in range(2, 10 // len(str_i) + 1):
            invalid_ids_set.add(int(str_i * r))

    invalid_ids = sorted(invalid_ids_set)
    pref_sum = [0, *accumulate(invalid_ids)]

    pairs = sys.stdin.read().strip().split(",")
    ans = 0
    for pair in pairs:
        l, r = map(int, pair.split("-"))
        pos_l = bisect_left(invalid_ids, l)
        pos_r = bisect_left(invalid_ids, r + 1)
        ans += pref_sum[pos_r] - pref_sum[pos_l]
    print(ans)


if __name__ == "__main__":
    main()

File: year/2025/day02/task2/main.cpp
/*
Author: Jose A. Romero
Puzzle: Advent of Code (year=2025 ; day=2 ; task=2)
*/

#include <bits/stdc++.h>

#define ll long long
#define endl '\n'

using namespace std;

int main() {
    ios_base::sync_with_stdio(0), cin.tie(0), cout.tie(0);

    set<ll> invalid_ids_set;
    for (int i = 0 ; i < 1e6 ; ++i) {
        string s = to_string(i), ss = s;

        for (int r = 10 / s.length() ; r >= 2 ; --r) {  // Number of repetitions, from 2 to 10 / s.length()
            ss += s;
            invalid_ids_set.insert(stoll(ss));
        }
    }
    vector<ll> invalid_ids(invalid_ids_set.begin(), invalid_ids_set.end());
    vector<ll> pref_sum = {0};
    for (auto x : invalid_ids) {
        pref_sum.push_back(pref_sum.back() + x);
    }

    string line;
    getline(cin, line);
    stringstream ss(line);

    ll l, r;
    char dash;
    ll ans = 0;
    while (ss >> l >> dash >> r) {
        int pos_l = lower_bound(invalid_ids.begin(), invalid_ids.end(), l) - invalid_ids.begin();
        int pos_r = lower_bound(invalid_ids.begin(), invalid_ids.end(), r + 1) - invalid_ids.begin();
        ans += pref_sum[pos_r] - pref_sum[pos_l];
        ss.ignore();  // Skip comma (or EOF)
    }
    cout << ans << endl;
    return 0;
}
